src/main/java/org/owasp/validator/html/Policy.java

```diff
@@ -79,957 +79,1031 @@
+
+	protected static ParseContext getSimpleParseContext(Element topLevelElement) throws PolicyException {
+		ParseContext parseContext = new ParseContext();
+		if (getByTagName(topLevelElement, "include").iterator().hasNext()) {
+			throw new IllegalArgumentException(
+					"A policy file loaded with an InputStream cannot contain include references");
+		}
+		parsePolicy(topLevelElement, parseContext);
+		return parseContext;
+	}
+
+	protected static ParseContext getParseContext(Element topLevelElement, URL baseUrl) throws PolicyException {
+		ParseContext parseContext = new ParseContext();
+
+		/**
+		 * Are there any included policies? These are parsed here first so that rules in
+		 * _this_ policy file will override included rules.
+		 *
+		 * NOTE that by this being here we only support one level of includes. To
+		 * support recursion, move this into the parsePolicy method.
+		 */
+		for (Element include : getByTagName(topLevelElement, "include")) {
+			String href = getAttributeValue(include, "href");
+
+			Element includedPolicy = getPolicy(href, baseUrl);
+			parsePolicy(includedPolicy, parseContext);
+		}
+
+		parsePolicy(topLevelElement, parseContext);
+		return parseContext;
+	}
+
+	protected static Element getTopLevelElement(final URL baseUrl) throws PolicyException {
+		final InputSource source = getSourceFromUrl(baseUrl);
+		return getTopLevelElement(source, new Callable<InputSource>() {
+			@Override
+			public InputSource call() throws PolicyException {
+				return getSourceFromUrl(baseUrl);
+			}
+		});
+	}
+
+	@SuppressFBWarnings(value = "SECURITY", justification = "Opening a stream to the provided URL is not "
+			+ "a vulnerability because it points to a local JAR file.")
+	protected static InputSource getSourceFromUrl(URL baseUrl) throws PolicyException {
+		try {
+			InputSource source = resolveEntity(baseUrl.toExternalForm(), baseUrl);
+			if (source == null) {
+				source = new InputSource(baseUrl.toExternalForm());
+				source.setByteStream(baseUrl.openStream());
+			} else {
+				source.setSystemId(baseUrl.toExternalForm());
+			}
+
+			return source;
+		} catch (SAXException | IOException e) {
+			// SAXException can't actually happen. See JavaDoc for resolveEntity(String,
+			// URL)
+			throw new PolicyException(e);
+		}
+	}
+
+	private static Element getTopLevelElement(InputStream is) throws PolicyException {
+		final InputSource source = new InputSource(is);
+		source.getByteStream().mark(0);
+		return getTopLevelElement(source, new Callable<InputSource>() {
+			@Override
+			public InputSource call() throws IOException {
+				source.getByteStream().reset();
+				return source;
+			}
+		});
+	}
+
+	protected static Element getTopLevelElement(InputSource source, Callable<InputSource> getResetSource)
+			throws PolicyException {
+		// Track whether an exception was ever thrown while processing policy file
+		Exception thrownException = null;
+		try {
+			return getDocumentElementFromSource(source, validateSchema);
```

Comment:
I think that altering this from true here will prevent the catch block content from 396 to 409 will never happen?
Can you verify that you still see the logger output expected from line 404 with these changes?

I believe @davewichers  wanted to maintain that workflow, so just want to be sure that still happens.

Reply:
If it's an malformed XML, it will be  thrown.
Example: 	
<tag-rules><css-rules></tag-rules></css-rules>

I'm adding that test case

Reply:
I was trying to work through it, and I'm more confident at this point that the `Invalid policy file` warning can never occur.

If we assume that the reference 'source' may be valid or invalid,  and validateSchema  may be true or false then there are four logical paths.

If I'm following the logic correctly, none of these paths can result in the `logger.warn("Invalid policy file: " + e.getMessage());` event or the element being returned from the second call to `getDocumentElementFromSource`
I've tried to outline each of the four paths below:


### 1 `validateSchema = true` and source is MALFORMED XML
Exception thrown from line 410

```
try {
    //throws SAXException since source is invalid XML
	return getDocumentElementFromSource(source, true);
} catch (SAXException e) {
	thrownException = e;
	//Skipped in this scenario, since validateSchema is true
	if (!validateSchema) {
		try {
			source = getResetSource.call();
			Element theElement = getDocumentElementFromSource(source, false);
			// We warn when the policy has an invalid schema, but schema validation is disabled.
			logger.warn("Invalid policy file: " + e.getMessage());
			return theElement;
		} catch (Exception e2) {
			throw new PolicyException(e2);
		}
	} else 
	    //  >> THIS THROW ESCAPES  when source is invalid and validateSchema is true
		throw new PolicyException(e);
} catch (ParserConfigurationException | IOException e) {
	thrownException = e;
	throw new PolicyException(e);
} finally {
	if (!validateSchema && (thrownException == null)) {
		// We warn when the policy has a valid schema, but schema validation is disabled.
		logger.warn("XML schema validation is disabled for a valid policy. Please reenable policy validation.");
	}
}
```

### 2 `validateSchema = false`  and source is MALFORMED XML
Exception thrown  from line 407
```
try {
    //throws SAXException since source is invalid XML
	return getDocumentElementFromSource(source, false);
} catch (SAXException e) {
	thrownException = e;
	if (!validateSchema) {
		try {
			source = getResetSource.call();
			//This is the exact same call as above.  SAXException is thrown.
			Element theElement = getDocumentElementFromSource(source, false);
			// We warn when the policy has an invalid schema, but schema validation is disabled.
			logger.warn("Invalid policy file: " + e.getMessage());
			return theElement;
		} catch (Exception e2) {
		    //  >> THIS THROW ESCAPES  when source is invalid and validateSchema is false
			throw new PolicyException(e2);
		}
	} else 
		throw new PolicyException(e);
} catch (ParserConfigurationException | IOException e) {
	thrownException = e;
	throw new PolicyException(e);
} finally {
	if (!validateSchema && (thrownException == null)) {
		// We warn when the policy has a valid schema, but schema validation is disabled.
		logger.warn("XML schema validation is disabled for a valid policy. Please reenable policy validation.");
	}
}
```


### 3  `validateSchema = true`  and source is WELL FORMED XML
Return value from line 395
```
try {
    //  >> RETURNS ELEMENT from here since everything is good with the xml.
	return getDocumentElementFromSource(source, true);
} catch (SAXException e) {
	thrownException = e;
	if (!validateSchema) {
		try {
			source = getResetSource.call();
			Element theElement = getDocumentElementFromSource(source, false);
			// We warn when the policy has an invalid schema, but schema validation is disabled.
			logger.warn("Invalid policy file: " + e.getMessage());
			return theElement;
		} catch (Exception e2) {
			throw new PolicyException(e2);
		}
	} else 
		throw new PolicyException(e);
} catch (ParserConfigurationException | IOException e) {
	thrownException = e;
	throw new PolicyException(e);
} finally {
	if (!validateSchema && (thrownException == null)) {
		// We warn when the policy has a valid schema, but schema validation is disabled.
		logger.warn("XML schema validation is disabled for a valid policy. Please reenable policy validation.");
	}
}
```

### 4 `validateSchema = false`   and source is WELL FORMED XML
Return value from line 395
```
try {
    //  >> RETURNS ELEMENT from here since everything is good with the xml.
	return getDocumentElementFromSource(source, false);
} catch (SAXException e) {
	thrownException = e;
	if (!validateSchema) {
		try {
			source = getResetSource.call();
			Element theElement = getDocumentElementFromSource(source, false);
			// We warn when the policy has an invalid schema, but schema validation is disabled.
			logger.warn("Invalid policy file: " + e.getMessage());
			return theElement;
		} catch (Exception e2) {
			throw new PolicyException(e2);
		}
	} else 
		throw new PolicyException(e);
} catch (ParserConfigurationException | IOException e) {
	thrownException = e;
	throw new PolicyException(e);
} finally {
	if (!validateSchema && (thrownException == null)) {
		// We warn when the policy has a valid schema, but schema validation is disabled.
		logger.warn("XML schema validation is disabled for a valid policy. Please reenable policy validation.");
	}
}
```

Unless I'm overlooking something, I think there will need to be further logic adjustments to achieve that output.